### ISSUE TITLE
causes 100% cpu Atom Helper issue

### ISSUE THREAD

**Issue by smnbbrv** [abe33/atom-pigments#125]
on MacBookPro on a specifically big project leads to 100% cpu issue. This appeared within some of the last updates, there was no problem like that before.

After a couple of hours this process becomes so annoying that the whole OS freezes and a laptop becomes so hot that I can cook pancakes.

Disabling the package solves the issue.

**Comment by ricardograca** [comment]
Same thing on Ubuntu. Here is a CPU profile to help diagnose the issue.
[CPU-20160108T120151.cpuprofile.zip](https://github.com/abe33/atom-pigments/files/82644/CPU-20160108T120151.cpuprofile.zip)

**Comment by abe33** [comment]
Hi guys, thanks for the report. 

@ricardograca thank you for providing a CPU profile, I'll take a look at it when back home. My guess is that there might be an infinite loop triggered somewhere in the background task (I encountered one once) but as it don't raise anything in the main thread it's quite hard to detect and debug.

**Comment by abe33** [comment]
I reviewed the code and found a condition leading to an infinite loop when parsing an expression, I published a patch for this case, let me know if you find some improvements with the new version.

**Comment by ricardograca** [comment]
That didn't fix the issue. I upgraded to 0.19.6, and as soon as I enabled the package again the CPU usage spiked again. It seems to be slightly lower than last time, but it's still very high. Do you want another profile?

**Comment by abe33** [comment]
@ricardograca Have you restarted Atom after the update? I'm not really sure how the compile cache plays for background tasks. 
Are you sure that's the Atom main process that reach 100% or is it the Atom helper one? I'm just trying to figure what could go wrong. 
Also, do you get the same behavior with a relatively small project ?

**Comment by ricardograca** [comment]
Yes, I've restarted it. I don't think it's the main process. The process that gets high CPU usage doesn't actually say "helper" anywhere, it's just a child process of the main one. With a small project the issue is not present.

I now waited about 10 minutes after enabling Pigments on a larger project and the atom process is still using a lot of CPU time:

![captura de ecra de 2016-01-12 11 09 15](https://cloud.githubusercontent.com/assets/1457035/12262081/61ce7f2e-b91d-11e5-9bbb-d1dafdd0c774.png)

Note that the RAM usage increases quickly to about 240MB and then immediately resets back to about 170MB, before increasing again, etc.

There is barely any I/O activity according to iotop.

**Comment by abe33** [comment]
By chance, is your project public? Or maybe is there a simple for me to reproduce the setup of your project?

**Comment by ricardograca** [comment]
No to both of those questions. It could be a particular file that is causing the problems right?

After some more testing I'm almost certain that this issue is not dependent on the project size. I think that it's something else, and bigger projects just have a higher chance of having that something that causes the issue.

I opened a few local forks of some relatively large projects (node-mapnik, mincer, bootstrap) and none showed high CPU usage with pigments enabled. I then opened my local fork of [bookshelf](https://github.com/tgriesser/bookshelf) and there was a constant ~20% CPU usage on one core. Lower than what I usually get on the other projects I'm working on but still significant.

Is there any chance that it could be related to git somehow? Because that's the only difference I see between all these projects. All the ones that exhibit the high CPU usage are things I have some work put into, whereas simple clones that I barely touched don't have this problem. This could also be a coincidence.

I'll see if I can isolate the issue to a particular file inside one of my affected projects.

**Comment by abe33** [comment]
>  It could be a particular file that is causing the problems right?

That's what I suspect, if threre's a file that contains some expressions that Pigments fails to parse, leading to an infinite loop somewhere (like when parsing expressions with nested parenthesis or square brackets) it could explain why the CPU is always this high (a background task is started, start looping over and over, but will be killed when a new task is started), that could also explain why some people don't see some colors, the failing task preventing every colors from the files further in the scan list from being properly scanned.

> After some more testing I'm almost certain that this issue is not dependent on the project size. I think that it's something else, and bigger projects just have a higher chance of having that something that causes the issue.

I completely agree, I've made a lot of improvements in the last month to prevent issues with big projects, and I never got this high CPU issue with projects like bootstrap or foundation.

> Is there any chance that it could be related to git somehow? 

It shouldn't, I work almost exclusively on projects with git support and never had any issues. Commits are stored as binary files in the .git directory which is automatically part of the VCS ignored files so they can't come up in pigments search.

> I'll see if I can isolate the issue to a particular file inside one of my affected projects.

That'll be great

**Comment by chodorowicz** [comment]
I have the same issue, some project cause 100% CPU usage, which goes away only by killing Atom. I will also try to isolate files since it's also not a public project.

**Comment by Mohamed-Abo-El-Soud** [comment]
just checking in, is there a fix for this?

**Comment by ricardograca** [comment]
Nope, mainly because no one knows what the problem is. If you want to help you can try to track which file in your project is causing the excessive CPU usage, and if possible, narrow it down even further.

**Comment by abe33** [comment]
@Mohamed-Abo-El-Soud It's just as @ricardograca said. actually I have no clue about what triggers that high CPU usage, I don't experience it myself (and I use Pigments intensively over a wide range of projects with various size and technologies) and have no means to reproduce the issue and thus find the cause and fix it. I'm really sorry that it's still open but as long as I can't reproduce that myself I fear I'll have to wait until someone can provide me with a sample project that triggers this issue or for someone to investigate the issue and provides a patch for that. 
As I already said, it might be just a single file that is causing the parser to enter a case I hadn't anticipated and make the worker loops indefinitely. I made a deep inspection previously and found some cases where it could occurs and fixed them, but I may have missed some. That's where I'm reduced to rely on you to help me on this.

**Comment by daveranan** [comment]
Same here, rMPB, 100% CPU usage. v0.24.5
I noticed the high CPU usage from pigments only after I had 5th project window opened.

**Comment by crysislinux** [comment]
Same here, cpu usage went to 100% when I opened the sourcecode of the newest extjs downloaded from https://www.sencha.com/products/extjs/#overview

**Comment by ricardograca** [comment]
"Me too" doesn't help this issue get fixed. If you want to help find the file or line that is causing the excessive CPU usage.

**Comment by Dayjo** [comment]
In response to https://github.com/abe33/atom-pigments/issues/169#issuecomment-203874894;

Thanks, unfortunately I cannot share the project as it's closed source. However, I have a variety of stylus,  and css files using sass syntax highlighting. I'll try and get a better idea of what in the project is causing the issue.

**Comment by abe33** [comment]
@Dayjo If you're using open source frameworks such compass or bootstrap, maybe you can list them here, I'll be able to a look at them to see if it comes from that.

**Comment by Dayjo** [comment]
Nothing like that I'm afraid. I have re-enabled it with just a single stylus file open which probably has 22 colours in it. It seems to be running fine so far. If I notice is spiking again i'll note what I've got open. Does it do anything with non-opened files in a project?

**Comment by abe33** [comment]
> Does it do anything with non-opened files in a project? 

Yep, that's how I can build the palette for the project and evaluate variables when used in other files. It first scans the project to find all the files that can be used as source for variables then it scans those files to find all the variables they defined. All this process being mostly done in a background task (except for open files, in that case I directly scans the buffer since it'll save me some IO)

**Comment by Dayjo** [comment]
Ok, well it's still running fine at the moment so it must be to do with a specific event / task I was doing!

**Comment by Dayjo** [comment]
So I tried to push it, I have maybe 15 `.styl` files open, some with no colours in, others with somewhere between 3 and 10 colours in.

With pigments enabled it's sitting at around 30% CPU, with it disabled at about 1%. 

Does it constantly scan open but inactive tabs, or does it just run the task on an 'on update' event or something? 

Also noticed this just when I went to look into one of the files - haven't seen this before and no idea if it's related;

![](http://c.dayjo.me/0y461a1p2W1j/Screen%20Recording%202016-03-31%20at%2005.47%20pm.gif)

I 'fixed' it by just clicking in there, writing a space, and then either 'undo' or deleting the space .. I think the thing I did is irrelevant, just that I changed the line. Could be a weird whitespace issue.

**Comment by rudolfratusinski** [comment]
I got this issue when i have open .ini files like php.ini. for each opened file like that i have 100% Usage (MBPro , El Capitan 10.11.4 atom-pigments 0.26 , atom 1.6.2)

**Comment by ricardograca** [comment]
@rudolfratusinski And can you share that file? Because without a sample file that consistently reproduces this issue it's not going to be fixed, and no one has provided a specific case or file that consistently reproduces this issue.

Just to be clear, if you open an empty project that has nothing except that `.ini` file you get consistently 100% CPU usage forever until you close Atom?

**Comment by rudolfratusinski** [comment]
Hi, Yes, of course.
I have some other addons, but all of them are disabled.

If i open one .ini file it takes 100% CPU. If i open two of them - 200% (in OSX activity monitor percentage is calculated per core, so 8 core PC can be stress-tested to 800%)

File is zipped because i couldn't upload an .ini file due to github restrictions.
In case it won't work on Your computer please let me know what data or logs You need.

[php.ini.zip](https://github.com/abe33/atom-pigments/files/216060/php.ini.zip)

**Comment by abe33** [comment]
@rudolfratusinski Thanks for the sample and the video, I'll take a look at that as soon as  come back home from work.

**Comment by abe33** [comment]
Hi did a quick test with the provided ini file, and while I can confirm that there's a worker that steadily stays above 90% of CPU, I don't reproduce that behaviour in dev mode (with my local pigments sources which is sync with current master) so I wonder if it's not related to the compile cache (see #95) as I suspect there's a difference between normal and dev mode in that regard. I'll take more time tonigh to figure out what can differ between the two mode.

**Comment by abe33** [comment]
Ok so, the good news is that I found why it failed with the ini file, and it should probably solve many other cases as well. 
Basically I wasn't properly scoping the expressions used to find variables in a file, so a scss or whatever could match some part of a file but failing to properly parse the match resulting in a process looping over and over on that part. Now I only use the expressions that was registered for the current file extension to avoid false matches. 

It'll take me some time before I can release the patch since I still get some weird failures locally and on TravisCI, and I also would like to fix the fact that an expression that was matched by regexp but not parsed properly leads to an infinite loop.

That's all for today 😄

**Comment by Dayjo** [comment]
Thanks for the hard work @abe33 :)

**Comment by houmark** [comment]
@abe33 Is this patch released yet? I once again have been forced to disable pigments due to 100% CPU. As soon as I disable pigments my CPU goes back to normal.

**Comment by abe33** [comment]
@houmark Yest I think it was released around version 0.26.0. But there might still some cases I haven't properly handled. As always, if you're able to isolate in a sample file/project a case that lead to these CPU spike feel free to post it here so that I can take a look at it.

**Comment by Dayjo** [comment]
@abe33 not had a problem since the last release.

**Comment by gugote** [comment]
Updated today to latest version and now its using 40% more or less, will keep testing it.

**Comment by Benjamski** [comment]
Great work on the update! I'm still getting cpu jumps in css files with many color codes but it is entirely usable. Will update if performance dips again below acceptable levels, but so far so good.

**Comment by fsworld009** [comment]
I also have this issue on Macbook Pro with 0.29.2

I'm not editing any file that has color codes when Atom Helper takes 100% cpu, all open tabs are React jsx files, plain js files, and one json file.

Once I disable Pigments the Atom Helper stops taking 100% cpu.

**Comment by ctcpip** [comment]
Anecdotal evidence based on some problems I was experiencing suggest that the cause may be due to large numbers of files that get read... see: `atom.packages.getActivePackage('pigments').mainModule.getProject().getPaths()`

One of the reasons these files can be so numerous is because default glob patterns for excluding files assume the project root will be the one containing `node_modules`, etc.

This may be better tracked in a separate issue, but I think default glob patterns may be better off like: `**/node_modules/*`

Also, consider excluding `bower_components` as well.  I'm curious why it isn't already there...

For me, I get around all this via project settings, but I think there would be value in updating these default patterns so people have a better experience out of the box.  That said, I may be missing some reason(s) why we wouldn't want to change the defaults as such.

**Comment by C-Bouthoorn** [comment]
Closing in favour of the [master post](https://github.com/abe33/atom-pigments/issues/351).
